Mathlib/LinearAlgebra/TensorAlgebra/Basic.lean

```diff
@@ -0,0 +1,325 @@
+        FreeAlgebra.hom_ext <|
+          funext fun x => by
+            rw [ι]
+            exact
+              (RingQuot.liftAlgHom_mkAlgHom_apply _ _ _ _).trans (FreeAlgebra.lift_ι_apply _ _) }
+#align tensor_algebra.lift TensorAlgebra.lift
+
+variable {R}
+
+@[simp]
+theorem ι_comp_lift {A : Type _} [Semiring A] [Algebra R A] (f : M →ₗ[R] A) :
+    (lift R f).toLinearMap.comp (ι R) = f := by
+  convert (lift R).symm_apply_apply f
+#align tensor_algebra.ι_comp_lift TensorAlgebra.ι_comp_lift
+
+@[simp]
+theorem lift_ι_apply {A : Type _} [Semiring A] [Algebra R A] (f : M →ₗ[R] A) (x) :
+    lift R f (ι R x) = f x := by
+  conv_rhs => rw [← ι_comp_lift f]
+#align tensor_algebra.lift_ι_apply TensorAlgebra.lift_ι_apply
+
+@[simp]
+theorem lift_unique {A : Type _} [Semiring A] [Algebra R A] (f : M →ₗ[R] A)
+    (g : TensorAlgebra R M →ₐ[R] A) : g.toLinearMap.comp (ι R) = f ↔ g = lift R f := by
+  rw [← (lift R).symm_apply_eq]
+  simp only [lift, Equiv.coe_fn_symm_mk]
+#align tensor_algebra.lift_unique TensorAlgebra.lift_unique
+
+-- Marking `TensorAlgebra` irreducible makes `Ring` instances inaccessible on quotients.
+-- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/algebra.2Esemiring_to_ring.20breaks.20semimodule.20typeclass.20lookup/near/212580241
+-- For now, we avoid this by not marking it irreducible.
+@[simp]
+theorem lift_comp_ι {A : Type _} [Semiring A] [Algebra R A] (g : TensorAlgebra R M →ₐ[R] A) :
+    lift R (g.toLinearMap.comp (ι R)) = g := by
+  rw [← lift_symm_apply]
+  exact (lift R).apply_symm_apply g
+#align tensor_algebra.lift_comp_ι TensorAlgebra.lift_comp_ι
+
+/-- See note [partially-applied ext lemmas]. -/
+@[ext]
+theorem hom_ext {A : Type _} [Semiring A] [Algebra R A] {f g : TensorAlgebra R M →ₐ[R] A}
+    (w : f.toLinearMap.comp (ι R) = g.toLinearMap.comp (ι R)) : f = g := by
+  rw [← lift_symm_apply, ← lift_symm_apply] at w
+  exact (lift R).symm.injective w
+#align tensor_algebra.hom_ext TensorAlgebra.hom_ext
+
+-- This proof closely follows `FreeAlgebra.induction`
+/-- If `C` holds for the `algebraMap` of `r : R` into `TensorAlgebra R M`, the `ι` of `x : M`,
+and is preserved under addition and muliplication, then it holds for all of `TensorAlgebra R M`.
+-/
+@[elab_as_elim]
+theorem induction {C : TensorAlgebra R M → Prop}
+    (h_grade0 : ∀ r, C (algebraMap R (TensorAlgebra R M) r)) (h_grade1 : ∀ x, C (ι R x))
+    (h_mul : ∀ a b, C a → C b → C (a * b)) (h_add : ∀ a b, C a → C b → C (a + b))
+    (a : TensorAlgebra R M) : C a := by
+  -- the arguments are enough to construct a subalgebra, and a mapping into it from M
+  let s : Subalgebra R (TensorAlgebra R M) :=
+    { carrier := C
+      mul_mem' := @h_mul
+      add_mem' := @h_add
+      algebraMap_mem' := h_grade0 }
+  -- porting note: split `let` in two to isolate a timeout in the second half
+  let of' : M →ₗ[R] (Subalgebra.toSubmodule s) := (ι R).codRestrict (Subalgebra.toSubmodule s) h_grade1
+  let of : M →ₗ[R] s := of'
```

Comment:
These were trivially defeq in mathlib3, but now aren't. The second `let` times out. I think this might again be due to new-style structures.